slm_lab/agent/algorithm/dqn.py

```diff
@@ -253,123 +235,85 @@ def post_body_init(self):
         )
         self.online_net = self.net
         self.eval_net = self.net
+        logger.info(util.self_desc(self))
 
-        # TODO handle this with better design
-        # create a new memory for task 2
-        MemoryClass = getattr(memory, _.get(self.agent.spec, 'memory.name'))
-        body_space = util.s_get(self, 'aeb_space.body_space')
-        env_bodies = body_space.get(e=0)
-        env_1_bodies = body_space.get(e=1)
-        self.agent.memory = MemoryClass(self.agent)
-        self.agent.memory.post_body_init(env_bodies)
-        self.agent.memory_1 = MemoryClass(self.agent)
-        self.agent.memory_1.post_body_init(env_1_bodies)
-
-    def get_batch(self):
-        batch_1 = self.agent.memory.get_batch(self.batch_size)
-        batch_2 = self.agent.memory_1.get_batch(self.batch_size)
-        # print("Inside get batch")
-        # print("Batch 1: ")
-        # print(batch_1)
-        # print("Batch 2: ")
-        # print(batch_2)
+    def sample(self):
+        # NOTE the purpose of multi-body is to parallelize and get more batch_sizes
+        batches = [body.memory.sample(self.batch_size)
+                   for body in self.agent.flat_nonan_body_a]
         # Package data into pytorch variables
-        float_data_list = [
-            'states', 'actions', 'rewards', 'dones', 'next_states']
-        for k in float_data_list:
-            batch_1[k] = Variable(torch.from_numpy(batch_1[k]).float())
-            batch_2[k] = Variable(torch.from_numpy(batch_2[k]).float())
+        for batch_b in batches:
+            util.to_torch_batch(batch_b)
         # Concat state
         combined_states = torch.cat(
-            [batch_1['states'], batch_2['states']], dim=1)
+            [batch_b['states'] for batch_b in batches], dim=1)
         combined_next_states = torch.cat(
-            [batch_1['next_states'], batch_2['next_states']], dim=1)
+            [batch_b['next_states'] for batch_b in batches], dim=1)
         batch = {'states': combined_states,
                  'next_states': combined_next_states}
         # use recursive packaging to carry sub data
-        batch['sub_1'] = batch_1
-        batch['sub_2'] = batch_2
+        batch['batches'] = batches
         return batch
 
     def compute_q_target_values(self, batch):
-        batch_1 = batch['sub_1']
-        batch_2 = batch['sub_2']
-        # print("batch: {}".format(batch['states'].size()))
-        # print("batch: {}".format(batch['states']))
-        q_vals = self.net.wrap_eval(batch['states'])
-        # Use act_select network to select actions in next state
-        # Depending on the algorithm this is either the current
-        # net or target net
-        q_next_st_act_vals = self.online_net.wrap_eval(
+        batches = batch['batches']
+        # TODO make split eval tail-wise softmax, otherwise will skew to just one tail
```

Comment:
@lgraesser a reason why agent succeeds only in once task, cuz softmax gives just 1 high prob value